Login: Fix wrong master bar displayed for non DOPS services

stephanethomas opened the pull request:
This pull request fixes a wrong master bar displayed for OAuth clients that are not one of our DOPS services (Akismet, Polldaddy, VaultPress ...), e.g.:

#### Before

<img width="624" alt="before" src="https://user-images.githubusercontent.com/594356/29974856-9153bc48-8f34-11e7-8d2e-5758f6a6b19e.png">

#### After

<img width="625" alt="after" src="https://user-images.githubusercontent.com/594356/29974865-9505b7a6-8f34-11e7-8381-31c91d9bf2fe.png">

Moreover, this pull request:

* Fixes a wrong master bar height for DOPS services
* Handles the new data structure returned by the API
* Enforces the logo size only from CSS
* Simplifies the code a bit, and adds a few more unit tests

It also fixes an unhandled promise rejection when the OAuth client data could not be retrieved from the API in order to display the appropriate error page. Finally, it makes sure we display the right subheader on the `Signup` page for all WooCommerce apps.

#### Testing instructions
 
1. Run `git checkout fix/oauth-master-bar` and start your server
2. Enable your proxy
3. Apply D7136-code patch to your sandbox
4. Point `public-api.wordpress.com` to your sandbox

##### Non DOPS service

5. Open the IntenseDebate [`Home` page](http://www.intensedebate.com/) in an incognito window
6. Click the `Sign In Here` link located in the upper right corner
7. Click the `Sign in with WordPress.com` button
8. Assert that you are redirected to the new [`Login` page](https://wordpress.com/log-in)
9. Duplicate your browser's current tab to ease comparison
10. Replace `https://wordpress.com` with `http://calypso.localhost:3000` in this new tab's url
11. Hit <kbd>Enter</kbd> to reload the page
12. Compare both pages and assert that the right master bar is now displayed

##### DOPS service with default look-and-feel

5. Open the Gravatar [`Home` page](https://fr.gravatar.com/) in French in an incognito window
6. Click the `Connexion` button in the upper right corner
7. Assert that you are redirected to the new [`Login` page](https://wordpress.com/log-in)
8. Duplicate your browser's current tab to ease comparison
9. Replace `https://wordpress.com` with `http://calypso.localhost:3000` in this new tab's url
10. Hit <kbd>Enter</kbd> to reload the page
11. Compare both pages and assert that the master bar now has the right height

##### DOPS service with custom look-and-feel

5. Open the WooCommerce [`Home` page](https://woocommerce.com/)
6. Click the `SIGN IN WITH WORDPRESS.COM` link
7. Assert that you are redirected to the new [`Login` page](https://wordpress.com/log-in)
8. Duplicate your browser's current tab to ease comparison
9. Replace `https://wordpress.com` with `http://calypso.localhost:3000` in this new tab's url
10. Hit <kbd>Enter</kbd> to reload the page
11. Compare both pages and assert that they look the same

#### Reviews
 
- [x] Code
- [x] Product
- [ ] Tests

jsnmoon commented on client/layout/logged-out.jsx:
```diff
@@ -32,17 +32,14 @@ const LayoutLoggedOut = ( {
-		if ( oauthClientName ) {
-			masterbar = <OauthClientMasterbar oauth2Client={ oauth2Client } />;
+		// Uses custom styles for DOPS clients - which are the only ones with the name property defined
+		if ( oauth2Client.name ) {
```
Would we want to use the conditional `( oauth2Client && oauth2Client.name )` here? The original logic accounted for `oauth2Client` being falsey (undefined, null, etc.).

pento replied:
No need, `useOAuth2Layout` is set by the `showOAuth2Layout()` selector, which does this test.

jsnmoon commented on client/layout/masterbar/oauth-client.jsx:
```diff
@@ -10,19 +10,19 @@ import PropTypes from 'prop-types';
 import { addLocaleToWpcomUrl, getLocaleSlug } from 'lib/i18n-utils';
 
-const OauthClientMasterbar = ( { oauth2Client } ) => (
+const OauthClientMasterbar = ( { oauth2Client } ) => (
```
Really minor: What happened here...?

pento replied:
It was previously a non-breaking space. `}` requires the Shift key to type, so the following space was probably typed before Shift was released. Very easy to do. :-)

jsnmoon commented on client/state/oauth2-clients/reducer.js:
```diff
@@ -94,22 +80,9 @@ const getClientIdFromSignupUrl = signupUrl => {
 export default createReducer( initialClientsData, {
 	[ OAUTH2_CLIENT_DATA_REQUEST_SUCCESS ]: ( state, { data } ) => {
-		const newData = Object.assign( {}, state[ data.id ], fromApi( data ) );
+		const newData = Object.assign( {}, state[ data.id ], data );
```
👍

jsnmoon commented on client/layout/masterbar/oauth-client.scss:
```diff
@@ -26,15 +25,19 @@
 }
 
+.masterbar__oauth-client-logo img {
+	max-height: 46px;
```
As is, max-height of 46px will put a blue border underneath the logo (due to `.masterbar`'s `height: 46px; border-bottom: 1px solid #0079aa;`). 

Should we use 47px instead to cover the blue bottom border?

(This is especially obvious with white background logos, like IntenseDebate.)

stephanethomas replied:
That's a good point indeed. I saw that problem during implementation and tried to set a `box-shadow` of `1px` at the bottom of the logo to be consistent with this blue border. Unfortunately, I couldn't find a way to make it work. I've just tried again to no avail, so I'll follow your suggestion for the time being. Maybe @danhauk can circle back on this when he has a few.

pento commented on client/signup/steps/user/index.jsx:
```diff
@@ -62,7 +62,7 @@ export class UserStep extends Component {
 			subHeaderText = '';
 		} else if ( flowName === 'wpcc' && oauth2Client ) {
-			if ( oauth2Client.id === 50916 ) {
+			if ( oauth2Client.name === 'woo' ) {
```
I think it's safer to stick with IDs, rather than compare it by name. Name currently isn't returned by the client-data endpoint, but I don't want to rely on that.

I'd prefer the failure state to be that we could accidentally show WordPress.com branding, rather than accidentally show Woo branding.

scruffian commented:
I think we should keep the default layout for non DOPS clients. Ideally in time we'll also remove the customizations for DOPS clients too, so that we have more consistency, but for now at least if we keep the default layout for other OAuth clients then we aren't introducing too much variation.

scruffian commented:
> I think we should keep the default layout for non DOPS clients. Ideally in time we'll also remove the customizations for DOPS clients too, so that we have more consistency, but for now at least if we keep the default layout for other OAuth clients then we aren't introducing too much variation.

I added bf956453ad to achieve this

pento commented on client/layout/logged-out.jsx:
```diff
@@ -31,18 +31,13 @@ const LayoutLoggedOut = ( {
-		classNameObject.dops = hasValidOAuth2ClientData;
-		classNameObject[ oauthClientName ] = hasValidOAuth2ClientData;
+	// Uses custom styles for DOPS clients - which are the only ones with the name property defined
+	if ( useOAuth2Layout && oauth2Client && oauth2Client.name ) {
```
`useOAuth2Layout` is only `true` if `oauth2Client` is defined, so the extra check isn't necessary.

stephanethomas replied:
That's what I thought as well but after giving it another look that's not entirely true.

`useOAuth2Layout()` will be `true` if `ui.oauth2Clients.currentClientId` exists, which happens as soon as we provide this id [in the url](https://github.com/Automattic/wp-calypso/blob/dd8c833c11d7b0c5fbb7ceedae6528a988269a63/client/state/ui/oauth2-clients/reducer.js#L15). It doesn't mean that the corresponding client data at `oauth2Clients` (in the global state tree) is available though (since we need to fetch it).

pento replied:
Oh, good catch! I totally missed that.

pento commented on client/lib/oauth2-clients.js:
```diff
@@ -0,0 +1,8 @@
+ */
+import { includes } from 'lodash';
+
+export const isWooOAuth2Client = ( oauth2Client ) => {
```
This function being in a completely different location to the `initialClientsData` blob in `oauth2-clients/reducer.js` will make it really hard to maintain.

stephanethomas replied:
@pento: I see your point, but on the other hand this function doesn't really belong to `/client/state` neither. What would you suggest?

pento replied:
Could `initialClientsData` be moved to `lib/oauth2-clients.js`? `oauth2-clients/reducer.js` could import it from there.

stephanethomas replied:
Well, we could move `initialClientsData` but to me it makes sense to keep it in `client/state`. Let's see what @Tug and @yurynix think of all of this.

yurynix replied:
I :heart: it!
Much better than `oauth2Client.name === 'woo'`.
Thank you for this.

re: `initialClientsData` :
* Do we even need `initialClientsData[].name` ? 
* Can we change implementation of `isWooOAuth2Client` to:
```
const isWooOAuth2Client = ( oauth2Client ) => !! oauth2Client.title.match(/WooCommerce/i);
```
From past mistakes, having hardcoded ids ( remember domain product ids in `/domains/` ? ) is a mess when couple of month later someone other than us ( @yoavf )  want to add a product and has no idea where / how / why. So IMO, we should avoid that at all cost.

* Can we add a comment to `oauth2-clients/reducer.js` saying:
> Hey... did you know? `client/lib/oauth2-clients.js` have some related stuff you should check out!

I don't think we should move initial state to here... but  ¯\_(ツ)_/¯ that's also a valid solution.

stephanethomas commented:
I've rebased, fixed merge conflicts, and edited Ben's commit to address [this comment](https://github.com/Automattic/wp-calypso/pull/17699#commitcomment-24075306).

scruffian commented:
Code 👍

scruffian commented:
Product 👍
